br/pkg/lightning/lightning.go

```diff
@@ -223,12 +225,70 @@ func (l *Lightning) RunServer() error {
 	}
 }
 
+// RunOnceWithOptions is used by binary lightning and host when using lightning as a library.
+// - for binary lightning, taskCtx could be context.Background which means taskCtx wouldn't be canceled directly by its
+//   cancel function, but only by Lightning.Stop or HTTP DELETE using l.cancel. No need to set Options
+// - for lightning as a library, taskCtx could be a meaningful context that get canceled outside, and there Options may
+//   be used:
+//   - WithGlue: set a caller implemented glue. Otherwise, lightning will use a default glue later.
+//   - WithDumpFileStorage: caller has opened an external storage for lightning. Otherwise, lightning will open a
+//     storage by config
+//   - WithCheckpointStorage: caller has opened an external storage for lightning and want to save checkpoint
+//     in it. Otherwise, lightning will save checkpoint by the Checkpoint.DSN in config
+func (l *Lightning) RunOnceWithOptions(taskCtx context.Context, taskCfg *config.Config, opts ...Option) error {
+	o := &options{}
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	failpoint.Inject("setExtStorage", func(val failpoint.Value) {
+		path := val.(string)
+		b, err := storage.ParseBackend(path, nil)
+		if err != nil {
+			panic(err)
+		}
+		s, err := storage.New(context.Background(), b, &storage.ExternalStorageOptions{})
+		if err != nil {
+			panic(err)
+		}
+		o.dumpFileStorage = s
+		o.checkpointStorage = s
+	})
+	failpoint.Inject("setCheckpointName", func(val failpoint.Value) {
+		file := val.(string)
+		o.checkpointName = file
+	})
+
+	// pre-check about options
+	// glue should be set when lightning in TiDB, and storages should be set when lightning in DM/dataflow engine,
+	// so they should not both be set.
```

Comment:
It seems glue and storage can both set technically.  If we implement  Lightning on SQL, TiDB also use Lightning as a library. At that time, TiDB may both set glue and storage.

Besides, Lightning is an independent component or library. All its interfaces should consider generic requirements not the one who depend on it.

Reply:
For lightning on SQL, it's not a requirement for TiDB to open an external storage. So if we emplace this restriction developers will use the lightning configuration to set loader dir and checkpoint DSN, which will reuse more logic about binary lightning and library lightning. In other words, my WithXXX options only want to make least changes to API.

Reply:
I would like that lightning binary and library use the same implementation to create the `RestoreController` later (This can be improved later).  WithXXX options indeed doesn't introduce much changes, but it is more like a hack that is similar to failpoint.

Reply:
> use the same implementation to create the RestoreController.

1. For binary, parse config file, create external storage and pass to `RestoreController`.
2. For library, pass external storage to `RestoreController` directly.